docs/design/assignment.md

```diff
@@ -0,0 +1,332 @@
+can only be used as complete statements, not as subexpressions of other
+operators, even when parenthesized:
+
+```
+var n: i32;
+// Error, assignment is not permitted as a subexpression.
+if (F() and (n = GetValue()) > 5) {
+}
+```
+
+User-defined types can define the meaning of these operations by
+[implementing an interface](#extensibility) provided as part of the Carbon
+standard library.
+
+## Syntax
+
+The operands of these operators can be any [expression](expressions/README.md).
+However, the first operand must be modifiable because it is passed to an
+`[addr self: Self*]` parameter, which disallows most expression forms other
+than:
+
+-   The name of a `var` binding.
+-   A dereference of a pointer.
+-   Array indexing that produces a modifiable result.
+-   Member access naming a field, where the object is one of these expressions.
+
+## Simple assignment semantics
+
+A simple assignment statement is intended to exactly mirror the semantics of
+initialization. The following two code snippets should have the same meaning if
+they are both valid:
+
+```
+// Declare and initialize.
+var v: T = init;
+```
+
+```
+// Declare separately from initialization.
+// Requires that `T` has an unformed state.
+var v: T;
+v = init;
+```
+
+This equivalence is not enforced, but when an object is in an unformed state,
+running the assignment function is _optional_, just like running the destructor
+is. If the assignment function is not run, the object will be directly
+initialized from the right-hand side instead. The type is still required to
+implement `AssignWith` for the assignment to be valid.
+
+```
+class C { ... }
+fn F() -> C {
+  returned var c: C = {...};
+  // `&c` here is `&x` for the first call to `F()`.
+  // `&c` here can be `&y` for the second call  to `F()`.
+  return var;
+}
+fn G() {
+  var x: C = F();
+  var y: C;
+  y = F();
+}
+```
+
+## Built-in types
+
+Integers and floating-point types, `bool`, and pointer types support simple
+assignment with `=`. The right-hand operand is implicitly converted to the type
+of the left-hand operand, and the converted value replaces the value of that
+operand.
+
+Compound assignment `$=` for integer and floating point types is
+[provided automatically](#defaults) for each supported operator `$`.
+
+For integer types, `++n;` and `--n;` behave the same as `n += 1;` and `n -= 1;`
+respectively. For floating-point types, these operators are not provided.
+
+## Tuples, structs, choice types, and data classes
+
+_TODO_: Describe the rules for assignment in these cases.
+
+See leads issue
+[#686: Operation order in struct/class assignment/initialization](https://github.com/carbon-language/carbon-lang/issues/686)
+
+## Extensibility
+
+Assignment operators can be provided for user-defined types by implementing the
+following families of interfaces. Implementations of these interfaces are
+provided for built-in types as necessary to give the semantics described above.
+
```

Comment:
I think it's worth documenting the design model we discussed today here -- here is my very rough sketch. Feel free to rewrite though, not sure if this is the best way of expressing it.

```suggestion
## Extensibility

Assignment operators can be provided for user-defined types by implementing the
following families of interfaces. Implementations of these interfaces are
provided for built-in types as necessary to give the semantics described above.

The *compound* assignment operators are expected to fundamentally be syntactic sugar in all cases, and default implementations of them will be synthesized from relevant assignment and operation implementations where available. There is no support for extending types to have compound assignment operators such as `a += b` that do not correspond to the relevant decomposed operators as in `a = a + b`.

In contrast, increment and decrement which are independently extensible. While for most numeric types, increment for example will map to `a = a + 1`, this is not universal or something Carbon defaults. Floating point types are one example where this doesn't hold true, as are types such as non-random-access iterators.
```

Reply:
Added something along these lines.